File: SegTree.py
from functools import reduce

class SegTree:
    '''
    通用线段树 by AK自动机
    支持增量更新，覆盖更新，序列更新，任意RMQ操作
    基于二叉树实现
    初始化：O(1)
    增量更新或覆盖更新的单次操作复杂度：O(log k)
    序列更新的单次复杂度：O(n)
    https://github.com/boristown/leetcode/blob/main/SegTree.py
    '''
    def __init__(self, f1, f2, l, r, v = 0):
        '''
        初始化线段树[left,right)
        f1,f2示例：
        线段和:
        f1=lambda a,b:a+b 
        f2=lambda a,n:a*n
        线段最大值:
        f1=lambda a,b:max(a,b)
        f2=lambda a,n:a
        线段最小值:
        f1=lambda a,b:min(a,b)
        f2=lambda a,n:a
        '''
        self.ans = '?'
        self.f1 = f1
        self.f2 = f2
        self.l = l #left
        self.r = r #right
        self.v = v #init value
        self.lazy_tag = 0 #Lazy tag
        self.left = None #SubTree(left,bottom)
        self.right = None #SubTree(right,bottom)
    
    @property
    def mid_h(self):
        return (self.l + self.r) // 2

    def create_subtrees(self):
        midh = self.mid_h
        if not self.left and midh > self.l:
            self.left = SegTree(self.f1, self.f2, self.l, midh)
        if not self.right:
            self.right = SegTree(self.f1, self.f2, midh, self.r)

    def init_seg(self, M):
        '''
        将线段树的值初始化为矩阵Matrx
        输入保证Matrx与线段大小一致
        '''
        m0 = M[0]
        self.lazy_tag = 0
        for a in M:
            if a!=m0:
                break
        else:
            self.v = m0
            return
        self.v = '#'
        self.ans = '?'
        midh = self.mid_h
        self.create_subtrees()
        self.left.init_seg(M[:midh-self.l])
        self.right.init_seg(M[midh-self.l:])
    
    def cover_seg(self, l, r, v):
        '''
        将线段[left,right)覆盖为val
        '''
        if self.v == v or l >= self.r or r <= self.l:
            return self.ans
        if l <= self.l and r >= self.r:
            self.v = v
            self.lazy_tag = 0
            self.ans = self.f2(v,self.r-self.l)
            return self.ans
        self.create_subtrees()
        if self.v != '#':
            if self.left:
                self.left.v = self.v
                self.left.ans = self.f2(v,self.left.r-self.left.l)
            if self.right:
                self.right.v = self.v
                self.right.ans = self.f2(v,self.right.r-self.right.l)
            self.v = '#'
        #push up
        self.ans = self.f1(self.left.cover_seg(l, r, v),self.right.cover_seg(l, r, v))
        return self.ans

    def inc_seg(self, l, r, v):
        '''
        将线段[left,right)增加val
        '''
        if v == 0 or l >= self.r or r <= self.l:
            return self.ans
        #self.ans = '?'
        if l <= self.l and r >= self.r:
            if self.v == '#':
                self.lazy_tag += v
            else:
                self.v += v
            self.ans += self.f2(v,self.left.r-self.left.l)
            return self.ans
        self.create_subtrees()
        if self.v != '#':
            self.left.v = self.v
            self.left.ans = self.f2(v,self.left.r-self.left.l)
            self.right.v = self.v
            self.right.ans = self.f2(v,self.right.r-self.right.l)
            self.v = '#'
        self.pushdown()
        self.ans = self.f1(self.left.inc_seg(l, r, v),self.right.inc_seg(l, r, v))
        return self.ans
        
    def pushdown(self):
        if self.lazy_tag != 0:
            if self.left:
                if self.left.v != '#':
                    self.left.v += self.lazy_tag
                    self.left.lazy_tag = 0
                else:
                    self.left.lazy_tag += self.lazy_tag
                self.left.ans += self.f2(self.lazy_tag, self.left.r-self.left.l)
            if self.right:
                if self.right.v != '#':
                    self.right.v += self.lazy_tag
                    self.right.lazy_tag = 0
                else:
                    self.right.lazy_tag += self.lazy_tag
                self.right.ans += self.f2(self.lazy_tag, self.right.r-self.right.l)
            self.lazy_tag = 0

    def query(self, l, r):
        '''
        查询线段[right,bottom)的RMQ
        '''
        if self.ans != '?' and l <= self.l and r >= self.r:
            return self.ans
        if self.v != '#':
            ans = self.f2(self.v, min(r,self.r)-max(l,self.l))
            if l <= self.l and r >= self.r:
                self.ans = ans
            return ans
        self.create_subtrees()
        midh = self.mid_h
        self.pushdown()
        anss = []
        if l < midh:
            anss.append(self.left.query(l, r))
        if r > midh:
            anss.append(self.right.query(l, r))
        ans = reduce(self.f1,anss)
        if l <= self.l and r >= self.r:
            self.ans = ans
        return ans

File: SegTree.cpp
#include <bits/stdc++.h>

using namespace std;

class SegTree{
    /*
    通用线段树 by AK自动机
    支持增量更新，覆盖更新，序列更新，任意RMQ操作
    基于二叉树实现
    初始化：O(1)
    增量更新或覆盖更新的单次操作复杂度：O(log k)
    序列更新的单次复杂度：O(n)
    https://github.com/boristown/leetcode/blob/main/SegTree.cpp
    */
   public:
        long _default,_ans;
        long (*_f1)(long,long),(*_f2)(long,long);
        long _l,_r,_v,_lazy_tag;
        SegTree *_left, *_right;

    SegTree(long (*f1)(long,long),long (*f2)(long,long), long l, long r, long v = 0){
        /*
        初始化线段树[left,right)
        f1,f2示例：
        线段和:
        long f1(long a,long b){
            return a+b;
        }
        long f2(long a,long n){
            return a*n;
        }
        线段最大值:
        long f1(long a,long b){
            return max(a,b);
        }
        long f2(long a,long n){
            return a;
        }
        线段最小值:
        long f1(long a,long b){
            return min(a,b);
        }
        long f2(long a,long n){
            return a;
        }
        */
        _default = v;
        _ans = (*f2)(v,r-l);
        _f1 = f1;
        _f2 = f2;
        _l = l; //left
        _r = r; //right
        _v = v; //init value
        _lazy_tag = 0; //Lazy tag
        _left = nullptr; //SubTree(left,bottom)
        _right = nullptr; //SubTree(right,bottom)
    }

    long get_mid(){
        return (_l + _r) / 2;
    }

    void create_subtrees(){
        long mid = get_mid();
        if(_left == nullptr && mid > _l)
            _left = new SegTree(_f1, _f2, _l, mid, _default);
        if(_right == nullptr)
            _right = new SegTree(_f1, _f2, mid, _r, _default);
    }

    long build(vector<long> arr){
        /*
        将线段树的值初始化为arr
        输入保证arr与线段大小一致
        */
        long m0 = arr[0];
        long len = arr.size();
        _lazy_tag = 0;
        if(_r == _l+1){
            _v = m0;
            _ans = (*_f2)(m0,arr.size());
            return _ans;
        }
        _v = -1;
        long mid = get_mid();
        create_subtrees();
        vector<long> arr_l;
        vector<long> arr_r;
        for(int i=0;i<len;i++){
            if(i<mid-_l)
                arr_l.emplace_back(arr[i]);
            else
                arr_r.emplace_back(arr[i]);
        }
        _ans = _f1(_left->build(arr_l), _right->build(arr_r));
        return _ans;
    }

    long cover_seg(long l,long r,long v){
        /*
        将线段[left,right)覆盖为val
        */
        if(_v == v || l >= _r || r <= _l) return _ans;
        if(l <= _l && r >= _r){
            _v = v;
            _lazy_tag = 0;
            _ans = _f2(v,_r-_l);
            return _ans;
        }
        create_subtrees();
        if(_v != -1){
            if(_left){
                _left->_v = _v;
                _left->_ans = _f2(_v,_left->_r - _left->_l);
                }
            if(_right){
                _right->_v = _v;
                _right->_ans = _f2(_v,_right->_r - _right->_l);
            }
            _v = -1;
        }
        //push up
        _ans = _f1(_left->cover_seg(l, r, v), _right->cover_seg(l, r, v));
        return _ans;
    }

    long inc_seg(long l,long r,long v){
        /*
        将线段[left,right)增加val
        */
        if(v == 0 || l >= _r || r <= _l) return _ans;
        //_ans = '?'
        if(l <= _l && r >= _r){
            if(_v == -1)
                _lazy_tag += v;
            else
                _v += v;
            _ans += _f2(v,_r-_l);
            return _ans;
        }
        create_subtrees();
        if(_v != -1){
            _left->_v = _v;
            _left->_ans = _f2(_v,_left->_r-_left->_l);
            _right->_v = _v;
            _right->_ans = _f2(_v,_right->_r-_right->_l);
            _v = -1;
        }
        pushdown();
        _ans = _f1(_left->inc_seg(l, r, v),_right->inc_seg(l, r, v));
        return _ans;
    }

    long inc_idx(long idx,long v){
        /*
        increase idx by val
        */
        if(v == 0 || idx >= _r || idx < _l)
            return _ans;
        if(idx == _l && _l == _r - 1){
            _v += v;
            _ans += _f2(v,1);
            return _ans;
        }
        create_subtrees();
        if(_v != -1){
            _left->_v = _v;
            _left->_ans = _f2(_v,_left->_r-_left->_l);
            _right->_v = _v;
            _right->_ans = _f2(_v,_right->_r-_right->_l);
            _v = -1;
        }
        pushdown();
        _ans = _f1(_left->inc_idx(idx, v),_right->inc_idx(idx, v));
        return _ans;
    }

    void pushdown(){
        if(_lazy_tag != 0){
            if(_left){
                if(_left->_v != -1){
                    _left->_v += _lazy_tag;
                    _left->_lazy_tag = 0;
                }
                else{
                    _left->_lazy_tag += _lazy_tag;
                }
                _left->_ans += _f2(_lazy_tag, _left->_r-_left->_l);
            }
            if(_right){
                if(_right->_v != -1){
                    _right->_v += _lazy_tag;
                    _right->_lazy_tag = 0;
                }
                else{
                    _right->_lazy_tag += _lazy_tag;
                }
                _right->_ans += _f2(_lazy_tag, _right->_r-_right->_l);
            }
            _lazy_tag = 0;
        }
    }

    long query(long l, long r){
        /*
        查询线段[right,bottom)的RMQ
        */
        if(l>=r) return 0;
        if(l <= _l && r >= _r)
            return _ans;
        if(_v != -1)
            return _f2(_v, min(_r, r) - max(_l, l));
        long mid = get_mid();
        vector<long> anss;
        if(l < mid)
            anss.emplace_back(_left->query(l, r));
        if(r > mid)
            anss.emplace_back(_right->query(l, r));
        long ans = 0;
        for(int i=0;i<anss.size();i++){
            if(i==0)
                ans = anss[i];
            else
                ans = _f1(ans,anss[i]);
        }
        return ans;
    }
};